### PR TITLE
MQTT5: support multiple Subscription ID properties

### PR DESCRIPTION
**Pull request by paul-lysak**
Motivation:

Subscription ID property of the PUBLISH message may be repeated multiple times, which wasn't taken into account when developing `MqttProperties` API.

Modification:

Store Subscription ID properties separately from others - in `MqttProperties.subscriptionIds`. 
Add `MqttProperties.getProperties` method to retrieve properties that may be repeated.
Change internal representation of User Properties for uniformity with Subscription ID - now they're stored in `MqttProperties.userProperties` rather than the common hash map.

Result:

Multiple Subscription ID properties can be set or retrieved.

### CODE REVIEW

**Comment by paul-lysak** [comment]
Also, now I think that introducing `MqttProperties.UserProperties` class to wrap multiple occurrences of `MqttProperties.UserProperty` was a bad idea. If `MqttProperties.listAll` returned a collection of `MqttProperty` instead, `MqttProperties.getProperty` returned first occurence of `UserProperty` and in order to retrieve multiple properties users would need to call `MqttProperties.getProperties` (introduced in this PR) we would have a cleaner API. 
However, the changes are already released and I'm not sure if it's OK to change the way properties behave now.

**Comment by normanmaurer** [comment]
@netty-bot test this please

@paul-lysak thats right we can not break the API anymore... This ship has sailed :/

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java [hunk 1]
```diff
@@ -139,20 +139,47 @@ public T value() {
...
 
     public static final class IntegerProperty extends MqttProperty<Integer> {
 
         public IntegerProperty(int propertyId, Integer value) {
             super(propertyId, value);
         }
+
+        @Override
+        public String toString() {
+            return "IntegerProperty(" + propertyId + "," + value + ")";
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java:169]: nit: `","` -> `", "` (add space)

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java [hunk 2]
```diff
@@ -200,26 +227,59 @@ public UserProperties(Collection<StringPair> values) {
             this.value.addAll(values);
         }
 
+        public static UserProperties fromUserPropertyCollection(Collection<UserProperty> properties) {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java:230]: does this need to be public ?

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java [hunk 3]
```diff
@@ -200,26 +227,59 @@ public UserProperties(Collection<StringPair> values) {
...
     }
 
     public static final class UserProperty extends MqttProperty<StringPair> {
         public UserProperty(String key, String value) {
             super(MqttPropertyType.USER_PROPERTY.value, new StringPair(key, value));
         }
+
+        @Override
+        public String toString() {
+            return "UserProperty(" + value.key + "," + value.value + ")";
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java:269]: nit: `","` -> `", "`

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java [hunk 4]
```diff
@@ -266,16 +337,79 @@ public void add(MqttProperty property) {
 
     public Collection<? extends MqttProperty> listAll() {
         IntObjectHashMap<MqttProperty> props = this.props;
-        return props == null? Collections.<MqttProperty>emptyList() : props.values();
+        if (props == null && subscriptionIds == null && userProperties == null) {
+            return Collections.<MqttProperty>emptyList();
+        } else if (subscriptionIds == null && userProperties == null) {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java:342]: nit: you can remove the `else` as you return

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java [hunk 5]
```diff
@@ -266,16 +337,79 @@ public void add(MqttProperty property) {
...
+            return props.values();
+        } else if (props == null && userProperties == null) {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java:344]: nit: you can remove the `else` as you return

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java [hunk 6]
```diff
@@ -266,16 +337,79 @@ public void add(MqttProperty property) {
...
+            return subscriptionIds;
+        } else {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java:346]: you can remove the `else` as you return before

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java [hunk 7]
```diff
@@ -266,16 +337,79 @@ public void add(MqttProperty property) {
...
+     */
     public MqttProperty getProperty(int propertyId) {
-        IntObjectHashMap<MqttProperty> props = this.props;
-        return props == null? null : props.get(propertyId);
+        if (propertyId == MqttPropertyType.USER_PROPERTY.value) {
+            //special handling to keep compatibility with earlier versions
+            List<UserProperty> userProperties = this.userProperties;
+            if (userProperties == null) {
+                return null;
+            } else {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java:379]: nit: you can remove the `else` as you return before

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java [hunk 8]
```diff
@@ -266,16 +337,79 @@ public void add(MqttProperty property) {
...
+                return UserProperties.fromUserPropertyCollection(userProperties);
+            }
+        } else if (propertyId == MqttPropertyType.SUBSCRIPTION_IDENTIFIER.value) {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java:382]: nit you can remove the `else`

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java [hunk 9]
```diff
@@ -266,16 +337,79 @@ public void add(MqttProperty property) {
...
+            List<IntegerProperty> subscriptionIds = this.subscriptionIds;
+            if (subscriptionIds == null || subscriptionIds.isEmpty()) {
+                return null;
+            } else {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java:386]: nit you can remove the `else`.

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java [hunk 10]
```diff
@@ -266,16 +337,79 @@ public void add(MqttProperty property) {
...
+     * Some properties (Subscription ID and User Properties) may occur few times,
+     * this method returns all their values in order.
+     *
+     * @param propertyId ID of the property
+     * @return all properties having specified ID
+     */
+    public List<? extends MqttProperty> getProperties(int propertyId) {
+        if (propertyId == MqttPropertyType.USER_PROPERTY.value) {
+            return userProperties == null ? Collections.<MqttProperty>emptyList() : userProperties;
+        } else if (propertyId == MqttPropertyType.SUBSCRIPTION_IDENTIFIER.value) {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java:406]: nit: you can remove the `else` as you return before

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java [hunk 11]
```diff
@@ -266,16 +337,79 @@ public void add(MqttProperty property) {
...
+            return subscriptionIds == null ? Collections.<MqttProperty>emptyList() : subscriptionIds;
+        } else {
```
Comment (normanmaurer) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java:408]: nit: you can remove the `else` as you return before

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java (hunk 2 above)
Comment (paul-lysak) [reply codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java]: Not really. Changed to `private`

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java [hunk 12]
```diff
@@ -266,16 +337,78 @@ public void add(MqttProperty property) {
...
+            }
+            return subscriptionIds.get(0);
+        }
+        IntObjectHashMap<MqttProperty> props = this.props;
+        return props == null ? null : props.get(propertyId);
+    }
+
+    /**
+     * Get properties by ID.
+     * Some properties (Subscription ID and User Properties) may occur few times,
```
Comment (chrisvest) [inline codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java:396]:
```suggestion
     * Some properties (Subscription ID and User Properties) may occur multiple times,
```

File: codec-mqtt/src/test/java/io/netty/handler/codec/mqtt/MqttPropertiesTest.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+                10,
+                ((MqttProperties.IntegerProperty) props.getProperty(SUBSCRIPTION_IDENTIFIER.value())).value.intValue());
+
+        List<MqttProperties.StringPair> expectedUserProps = new ArrayList<MqttProperties.StringPair>();
+        expectedUserProps.add(new MqttProperties.StringPair("isSecret", "true"));
+        expectedUserProps.add(new MqttProperties.StringPair("tag", "firstTag"));
+        expectedUserProps.add(new MqttProperties.StringPair("tag", "secondTag"));
+        List<MqttProperties.StringPair> actualUserProps =
+                ((MqttProperties.UserProperties) props.getProperty(USER_PROPERTY.value())).value;
+        assertEquals("getProperty User Properties", actualUserProps, expectedUserProps);
```
Comment (chrisvest) [inline codec-mqtt/src/test/java/io/netty/handler/codec/mqtt/MqttPropertiesTest.java:62]: The expected value goes before the actual value in `assertEquals`

File: codec-mqtt/src/test/java/io/netty/handler/codec/mqtt/MqttPropertiesTest.java [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+                expectedSubscriptionIds,
+               props.getProperties(SUBSCRIPTION_IDENTIFIER.value()));
+
+        List<MqttProperties.UserProperty> expectedUserProps = new ArrayList<MqttProperties.UserProperty>();
+        expectedUserProps.add(new MqttProperties.UserProperty("isSecret", "true"));
+        expectedUserProps.add(new MqttProperties.UserProperty("tag", "firstTag"));
+        expectedUserProps.add(new MqttProperties.UserProperty("tag", "secondTag"));
+        List<MqttProperties.UserProperty> actualUserProps =
+                (List<MqttProperties.UserProperty>) props.getProperties(USER_PROPERTY.value());
+        assertEquals("getProperty User Properties", actualUserProps, expectedUserProps);
```
Comment (chrisvest) [inline codec-mqtt/src/test/java/io/netty/handler/codec/mqtt/MqttPropertiesTest.java:86]: 📌

**Review by chrisvest — COMMENTED** [review]
Some small nits, but looks fine.

**Review by chrisvest — APPROVED** [review]

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Review by normanmaurer — APPROVED** [review]

**Pull request merged**